pkg/k8s/cnpstatus.go

```diff
@@ -0,0 +1,345 @@
+}
+
+func newCNPEventMap() *cnpEventMap {
+	return &cnpEventMap{
+		eventMap: make(map[string]*NodeStatusUpdater),
+	}
+}
+
+func (c *cnpEventMap) lookup(cnpKey string) (*NodeStatusUpdater, bool) {
+	c.RLock()
+	ch, ok := c.eventMap[cnpKey]
+	c.RUnlock()
+	return ch, ok
+}
+
+func (c *cnpEventMap) createIfNotExist(cnpKey string) (*NodeStatusUpdater, bool) {
+	c.Lock()
+	nsu, ok := c.eventMap[cnpKey]
+	// Cannot reinsert into map when active channel present.
+	if ok {
+
+		c.Unlock()
+		return nsu, ok
+	}
+	nsu = &NodeStatusUpdater{
+		updateChan: make(chan *NodeStatusUpdate, 512),
+		stopChan:   make(chan struct{}),
+	}
+	c.eventMap[cnpKey] = nsu
+	c.Unlock()
+	return nsu, ok
+}
+
+func (c *cnpEventMap) delete(cnpKey string) {
+	c.Lock()
+	nsu, ok := c.eventMap[cnpKey]
+	if !ok {
+		return
+	}
+	// Signal that we should stop processing events.
+	close(nsu.stopChan)
+	delete(c.eventMap, cnpKey)
+	c.Unlock()
+}
+
+// NewCNPStatusEventHandler returns a new CNPStatusEventHandler.
+func NewCNPStatusEventHandler(cnpStore *store.SharedStore, k8sStore cache.Store, updateInterval time.Duration) *CNPStatusEventHandler {
+	return &CNPStatusEventHandler{
+		eventMap:       newCNPEventMap(),
+		cnpStore:       cnpStore,
+		k8sStore:       k8sStore,
+		updateInterval: updateInterval,
+	}
+}
+
+// NodeStatusUpdate pairs a CiliumNetworkPolicyNodeStatus to a specific node.
+type NodeStatusUpdate struct {
+	node string
+	*cilium_v2.CiliumNetworkPolicyNodeStatus
+}
+
+// WatchForCNPStatusEvents starts a watcher for all CNP status updates from
+// the key-value store.
+func (c *CNPStatusEventHandler) WatchForCNPStatusEvents() {
+
+restart:
+	watcher := kvstore.Client().ListAndWatch("cnpStatusWatcher", CNPStatusesPath, 512)
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				log.Debugf("%s closed, restarting watch", watcher.String())
+				time.Sleep(500 * time.Millisecond)
+				goto restart
+			}
+
+			switch event.Typ {
+			case kvstore.EventTypeListDone, kvstore.EventTypeDelete:
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				var cnpStatusUpdate CNPNSWithMeta
+				err := json.Unmarshal(event.Value, &cnpStatusUpdate)
+				if err != nil {
+					log.WithFields(logrus.Fields{"kvstore-event": event.Typ.String(), "key": event.Key}).
+						WithError(err).Error("Not updating CNP Status; error unmarshaling data from key-value store")
+					continue
+				}
+
+				log.WithFields(logrus.Fields{
+					"uid":       cnpStatusUpdate.UID,
+					"name":      cnpStatusUpdate.Name,
+					"namespace": cnpStatusUpdate.Namespace,
+					"node":      cnpStatusUpdate.Node,
+					"key":       event.Key,
+					"type":      event.Typ,
+				}).Debug("received event from kvstore")
+
+				// Send the update to the corresponding controller for the
+				// CNP which sends all status updates to the K8s apiserver.
+				cnpKey := fmt.Sprintf("%s/%s/%s", cnpStatusUpdate.UID, cnpStatusUpdate.Namespace, cnpStatusUpdate.Name)
```

Comment:
I think it's better to create a method that generates the key. I've seen similar code in different places which can be error prone. Besides, this key is different than the one in the commit message:

```
cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>/<node name> -
```

Reply:
Agreed re: having a function. However, the commit message refers to the format in the key-value store. The format in the key you refer to is not for the key-value store.